src/java.base/share/native/libjli/java.c

```diff
@@ -534,6 +582,329 @@ invokeInstanceMainWithoutArgs(JNIEnv *env, jclass mainClass) {
+ * @return EXECVP_SUCCESS if the execution of the 'criu restore' command succeeds, error code otherwise
+ */
+static int
+restoreFromCheckpoint(const char *checkpointDirectory, const char *logLevel, jboolean unprivilegedModeOn, const char *logFile)
+{
+    int restoreStatus = EXECVP_SUCCESS;
+    int length = -1;
+    char *logLevelOption = NULL;
+    char *logFileOption = NULL;
+    int argc = 0;
+    const char *argv[9] = { NULL };
+    argv[argc++] = CRIU_COMMAND;
+    argv[argc++] = CRIU_RESTORE_OPTION;
+    argv[argc++] = CRIU_CHECKPOINT_DIRECTORY_OPTION;
+    argv[argc++] = checkpointDirectory;
+    if (NULL == logLevel) {
+        argv[argc++] = CRIU_DEFAULT_LOG_LEVEL_OPTION;
+    } else {
+        length = snprintf(NULL, 0, LOG_LEVEL_OPTION_FORMAT, atoi(logLevel));
+        if (length < 0) {
+            JLI_ReportErrorMessage(COMMAND_OPTION_LENGTH_CALCULATION_ERROR_MESSAGE, logLevel, LOG_LEVEL_OPTION_FORMAT);
+            restoreStatus = MEMORY_ALLOCATION_ERROR;
+            goto done;
+        }
+        logLevelOption = (char *)JLI_MemAlloc(length + 1);
+        if (NULL == logLevelOption) {
+            JLI_ReportErrorMessage(COMMAND_OPTION_MEMORY_ALLOCATION_ERROR_MESSAGE, logLevel, LOG_LEVEL_OPTION_FORMAT);
+            restoreStatus = MEMORY_ALLOCATION_ERROR;
+            goto done;
+        }
+        if (snprintf(logLevelOption, length + 1, LOG_LEVEL_OPTION_FORMAT, atoi(logLevel)) < 0) {
+            JLI_ReportErrorMessage(COMMAND_OPTION_MEMORY_ALLOCATION_ERROR_MESSAGE, logLevel, LOG_LEVEL_OPTION_FORMAT);
+            restoreStatus = MEMORY_ALLOCATION_ERROR;
+            goto freeLogLevelOption;
+        }
+        argv[argc++] = logLevelOption;
+    }
+    argv[argc++] = CRIU_SHELL_JOB_OPTION;
+    if (unprivilegedModeOn) {
+        argv[argc++] = CRIU_UNPRIVILEGED_MODE_OPTION;
+    }
+    if (NULL != logFile) {
+        length = snprintf(NULL, 0, LOG_FILE_OPTION_FORMAT, logFile);
+        if (length < 0) {
+            JLI_ReportErrorMessage(COMMAND_OPTION_LENGTH_CALCULATION_ERROR_MESSAGE, logFile, LOG_FILE_OPTION_FORMAT);
+            restoreStatus = MEMORY_ALLOCATION_ERROR;
+            goto freeLogLevelOption;
+        }
+        logFileOption = (char *)JLI_MemAlloc(length + 1);
+        if (NULL == logFileOption) {
+            JLI_ReportErrorMessage(COMMAND_OPTION_MEMORY_ALLOCATION_ERROR_MESSAGE, logFile, LOG_FILE_OPTION_FORMAT);
+            restoreStatus = MEMORY_ALLOCATION_ERROR;
+            goto freeLogLevelOption;
+        }
+        if (snprintf(logFileOption, length + 1, LOG_FILE_OPTION_FORMAT, logFile) < 0) {
+            JLI_ReportErrorMessage(COMMAND_OPTION_MEMORY_ALLOCATION_ERROR_MESSAGE, logFile, LOG_FILE_OPTION_FORMAT);
+            restoreStatus = MEMORY_ALLOCATION_ERROR;
+            goto freeLogFileOption;
+        }
+        argv[argc++] = logFileOption;
+    }
+    argv[argc] = NULL;
+    if (EXECVP_ERROR == execvp(argv[0], (char * const *)argv)) {
+        restoreStatus = EXECVP_ERROR;
+    }
```

Comment:
It's not clear there's any value in checking the return value from `execvp()` - it doesn't return when successful.